File: Arboles/bfs/bfs-lista-ady.py
from sys import stdin
from collections import deque

MAX, n = 10000, int()
adj = [[] for _ in range(MAX)]
visitados = [False for _ in range(MAX)]
#p = [-1 for _ in range(MAX)]

def bfsAux(u):
    cola = deque()
    cola.append(u)
    visitados[u] = True

    while len(cola) != 0:
        w = cola.popleft()
        print(w)

        for v in adj[w]:
        #for i in range(len(adj[v])):
            #w = adj[w][i]
            if not visitados[v]:
                cola.append(v)
                visitados[v] = True
                #p[v] = v

def bfs():
    for i in range(n):
        visitados[i] = False
        #p[i] = -1
    for i in range(n):
        if not visitados[i]:
            bfsAux(i)

def main():
    global n, visitados, adj
    n, m = list(map(int, stdin.readline().split()))

    for i in range(n):
        vals = list(map(int, stdin.readline().split()))
        u = vals[0]
        for j in range(u):
            adj[i].append(vals[j + 1])

    print("Grafo")
    for i in range(n):
        print("Nodo %d:" % i)
        for j in range(len(adj[i])):
            print("%d" % adj[i][j], end = ' ')
        print("")

    print("BFS")
    bfs()
    for i in range(n):
        print(p[i], end = ' ')
    print("")

main()

File: Arboles/bfs/bfs-lista-ady.cpp
#include <vector>
#include <map>
#include <string>
#include <cstdio>
#include <iostream>
#include <queue>

using namespace std;

int n;
vector<vector<int>> adj(10000);
bool visitados[1000];
int p[1000];

void bfsAux(int v){
  int u, w, i;
  queue<int> cola;
  visitados[v] = true;
  cola.push(v);

  while(!cola.empty()){
    w = cola.front();
    printf("%d\n", w);
    cola.pop();

    for(i = 0; i < adj[w].size(); ++i){
      v = adj[w][i];
      if(!visitados[v]){
	visitados[v] = true;
	//p[v] = u;
	cola.push(v);
      }
    }
  }
}

void bfs(){
  int i;

  //fill(visitados, visitados + n, false);
  for(i = 0; i < n; ++i){
    visitados[i] = false;
    //p[i] = -1;
  }

  for(i = 0; i < n; ++i){
    if(!visitados[i])
      bfsAux(i);
  }
}

int main(){
  int m, u, v, i, j;

  cin >> n >> m;

  for(i = 0; i < n; ++i){
    cin >> u;
    adj[i].resize(u);
    for(j = 0; j < u; ++j)
      cin >> adj[i][j];
  }

  cout << "Grafo" << endl;

  for(i = 0; i < n; ++i){
    cout << "Nodo " << i << ":";
    for(j = 0; j < adj[i].size(); ++j)
      cout << " " << adj[i][j];
    cout << endl;
  }

  cout << "BFS:" << endl;
  bfs();

  return 0;
}
